src/librustc_typeck/collect.rs

```diff
@@ -1385,7 +1385,7 @@ fn predicates_of<'a, 'tcx>(tcx: TyCtxt<'a, 'tcx, 'tcx>,
                 }
             };
 
-            let assoc_ty = tcx.mk_projection(self_trait_ref, trait_item.name);
+            let assoc_ty = tcx.mk_deprecated_projection(self_trait_ref, trait_item.name);
```

Comment:
You can turn `trait_item_ref.id` into a DefId here - and you get the associated type DefId.